### PR TITLE
[CUDAGraph] Graph Partition

### PR DESCRIPTION
**Pull request by BoyuanFeng**
This PR implements cudagraph partition, following previous PR on inductor graph partition (#147038). Since there are many ops that cudagraph cannot support, this PR focuses on `cpu ops` and will add more partition rules in the next PR.

## Example
```python
import torch

torch._inductor.config.graph_partition = True


def f(x, y):
    x1 = x + 1
    y1 = y + 1
    y_cpu = y1.cpu() + 1
    z = x @ y
    return x1 + y1 + z + y_cpu.cuda()


x, y = [torch.ones(2, 2, device="cuda") for _ in range(2)]
x_cloned, y_cloned = [tmp.clone() for tmp in [x,y]]
eager_out = f(x, y)


f_compiled = torch.compile(f, mode="reduce-overhead")


for _ in range(5):
    compiled_out = f_compiled(x_cloned, y_cloned)
    assert torch.allclose(eager_out, compiled_out)
```

w/o graph partition, we will skip cudagraph:
```
skipping cudagraphs due to skipping cudagraphs due to cpu device (device_put). Found from : 
   File "/home/boyuan/playground/cudagraph/graph_partition/graph_partition.py", line 9, in f
    y_cpu = y1.cpu() + 1 # 3
```

w/ graph partition, we can see two cudagraphify under the same torch-compiled region:
image

## Design

PR #147038 splits `def call(args)` function into multiple `def partition_id(args)`. In this PR, we use `recursively_apply_fns()` to wrap each `partition_id()` function with `cudagraphify`. One major design point is, `cudagraphify` takes metadata such as static_input_idxs and we need to provide such metadata for each graph partition. However, we previously only have such metadata for the original graph instead of graph partitions.

The [idea](https://github.com/pytorch/pytorch/pull/147038#discussion_r1964124800) is: 
- compute a mapping from the partition metadata (e.g., input/output idx) to the graph metadata, stored in `GraphPartitionMap`. 
- during post_compile, get the `CudagraphMetadata` for each partition based on the graph-level metadata and `GraphPartitionMap`, via `get_partition_cudagraph_metadata()`.
- finally, in `cudagraph_partition_pos_compile`, we compute the `CudagraphMetadata` and apply cudagraphify for each graph via `recursively_apply_fns`.

#### Q: How does it work with codecache?

While we have multiple graph partitions, we still have 1 file and 1 `call` function for 1 dynamo graph. The major difference is we need to additionally load a `recursively_apply_fns()` for graph partition. We also add `partition_maps: Optional[list[GraphPartitionMap]]` to `CompiledFxGraph` so it will be serialized and could be deserialized later.

## Edge Case 1
PyTorch has an assumption on input/output orders. For example, backward inputs take saved tensors first and then tangents. In graph partition, we respect such orders via `graph_partition_signature_reorder`.

## Edge Case 2
Cudagraphifying `call` function gives 2 cudagraph managed tensors `buf0` and `primals_1`. However, cudagraphifying `partition_0` gives only 1 cudagraph managed tensor `buf0`. This leads to a semantic difference between cudagraph w/ and w/o graph partition. [full code comparison](https://www.internalfb.com/intern/diffing/?paste_number=1747654420)

image

To achieve the same semantic, we returns an input tensor as output if it is not freed in a graph partition. This allows more cudagraph managed tensors and is important for handling saved tensors.

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @ekr0 @xmfan @eellison @zou3519 @jamesjwu

### CODE REVIEW

File: torch/_inductor/output_code.py [hunk 1]
```diff
@@ -132,6 +136,48 @@ def complex_memory_overlap(t: torch.Tensor) -> bool:
     return False
 
 
+def try_handle_backward_generation(compiled_graph: CompiledFxGraph) -> None:
```
Comment (eellison) [inline torch/_inductor/output_code.py:139]: try makes me a bit nervous :nervous - we should always be able to handle this. could you rename ? if the try is to indicate we don't always do it, you could do `maybe` instead

File: torch/_inductor/scheduler.py [hunk 2]
```diff
@@ -3979,9 +3983,13 @@ def should_partition(self, node: BaseSchedulerNode) -> bool:
...
-            isinstance(expr, sympy.Expr) and expr.free_symbols
-            for expr in node.node.layout.size
+        if (
+            hasattr(node.node, "layout")
+            and hasattr(node.node.layout, "size")
+            and any(
+                isinstance(expr, sympy.Expr) and expr.free_symbols
+                for expr in node.node.layout.size
+            )
         ):
```
Comment (eellison) [inline torch/_inductor/scheduler.py:3993]: I guess this is to be handled in a follow up.

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -4003,6 +4011,69 @@ def get_name_to_nodes(
...
+            )
+
+    def graph_partition_signature_reorder(
+        self, names: OrderedSet[str], order: OrderedSet[str]
+    ) -> OrderedSet[str]:
+        """
+        Returns an ordered set of names that includes all strings in 'names'
+        but follows the order in 'order'. This supports the assumption on input
+        and output orders, such as backward inputs take saved tensors first and
+        then tangents.
```
Comment (eellison) [inline torch/_inductor/scheduler.py:4063]: I dont follow this. 

> This supports the assumption on input and output orders, such as backward inputs take saved tensors first and then tangents.

What is breaking currently ?

**Review by eellison — COMMENTED** [review]
looking good, a few questions

File: test/inductor/test_cudagraph_trees.py [hunk 4]
```diff
@@ -2510,6 +2510,187 @@ def f(x):
...
+            with capture_stderr() as captured_output:
+                for _ in range(3):
+                    foo(
+                        torch.ones([10], device="cuda:0"),
+                        torch.ones([10], device="cuda:1"),
+                    )
+
+            FileCheck().check("skipping cudagraphs due to multiple devices").run(
+                captured_output[0]
+            )
```
Comment (eellison) [inline test/inductor/test_cudagraph_trees.py:2689]: > Cudagraphifying call function gives 2 cudagraph managed tensors buf0 and primals_1. However, cudagraphifying partition_0 gives only 1 cudagraph managed tensor buf0. This leads to a semantic difference between cudagraph w/ and w/o graph partition. [full code comparison](https://www.internalfb.com/intern/diffing/?paste_number=1747654420)

Can you comment more on this ? im not sure i follow.

File: torch/_inductor/utils.py [hunk 5]
```diff
@@ -151,6 +151,24 @@ def eval(cls, value: sympy.Expr) -> Optional[sympy.Expr]:
...
+    """
+    Mapping from the partition info (e.g., input/output) to the graph info
+    """
+
+    # a unique id of graph partition
+    id: int
+
+    # map partition input/output indices to graph input/output indices. None indicates
+    # a partition input/output is not a graph input/output.
+    input_index_mapping: list[Optional[int]]
```
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/utils.py:165]: when is it multiple values ?

File: torch/_inductor/scheduler.py [hunk 6]
```diff
@@ -4049,15 +4099,34 @@ def get_graph_partition_signature(
...
-                GraphPartitionSignature(
-                    input_nodes,
-                    output_nodes,
-                    input_deallocation,
-                    skip_cudagraph,
-                )
+
+            constant_names = [
+                name for name in partition_input_names if name not in name_to_node
+            ]
```
Comment (eellison) [inline torch/_inductor/scheduler.py:4118]: This seems like a weird way of getting constants.. is there another way of doing this ? or could you say more about this

File: torch/_inductor/scheduler.py [hunk 7]
```diff
@@ -4049,15 +4099,34 @@ def get_graph_partition_signature(
                 for name in partition_input_names
                 if name in name_to_node
             }
+
+            # returns an input tensor as output if it is not freed. This allows
+            # more cudagraph managed tensors and is important for handling saved
+            # tensors.
```
Comment (eellison) [inline torch/_inductor/scheduler.py:4105]: add a bit more comment on this ? i didnt follow on first pass.

File: torch/_inductor/utils.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/utils.py:165]: When  would this be `[None, 3]` for instance ?

**Review by eellison — APPROVED** [review]

File: torch/_inductor/scheduler.py (hunk 6 above)
Comment (eellison) [reply torch/_inductor/scheduler.py:4118]: Could we use constant names instead ?
